onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc

```diff
@@ -0,0 +1,476 @@
+    auto index = AddNewNNAPIOperand(operandType);                             \
+    THROW_ON_ERROR_WITH_NOTE(                                                 \
+        nnapi_->ANeuralNetworksModel_setOperandValue(                         \
+            nnapi_model_->model_, index, &value, sizeof(value)),              \
+        "value: " + std::to_string(value));                                   \
+    return index;                                                             \
+  }
+
+DEFINE_ADD_OPERAND_FROM_SCALAR(bool, BOOL);
+DEFINE_ADD_OPERAND_FROM_SCALAR(int32_t, INT32);
+DEFINE_ADD_OPERAND_FROM_SCALAR(float, FLOAT32);
+
+#undef DEFINE_ADD_OPERAND_FROM_SCALAR
+
+void ModelBuilder::AddInitializerToSkip(const std::string& tensor_name) {
+  skipped_initializers_.insert(tensor_name);
+}
+
+void ModelBuilder::Prepare() {
+  nnapi_model_ = std::unique_ptr<Model>(new Model());
+  THROW_ON_ERROR(nnapi_->ANeuralNetworksModel_create(&nnapi_model_->model_));
+  PreprocessIntializers();
+  RegisterInitializers();
+  RegisterModelInputs();
+  AddOperations();
+  RegisterModelOutputs();
+  RegisterModelShaper();
+}
+
+constexpr size_t kDefaultByteAlignmentForNNAPI = 16;
+static size_t GetPaddedByteSize(size_t size) {
+  // if (size_t r = size % kDefaultByteAlignmentForNNAPI)
+  //   return size + kDefaultByteAlignmentForNNAPI - r;
+  // else
+  //   return size;
+  // This does exactly the same as the logic above
+  return (size + kDefaultByteAlignmentForNNAPI - 1) & ~(kDefaultByteAlignmentForNNAPI - 1);
+}
+
+void ModelBuilder::GetAllInitializers() {
+  for (const auto& tensor : model_proto_.graph().initializer()) {
+    initializers_.insert({tensor.name(), tensor});
+  }
+}
+
+void ModelBuilder::PreprocessIntializers() {
+  for (const auto& node : model_proto_.graph().node()) {
+    if (auto* opBuilder = GetOpBuilder(node)) {
+      opBuilder->AddInitializersToSkip(*this, node);
+    }
+  }
+}
+
+void ModelBuilder::RegisterInitializers() {
+  // First pass to get all the stats of the initializers
+  auto initializer_size = model_proto_.graph().initializer_size();
+  std::vector<std::tuple<uint32_t, size_t, size_t>> initializers(initializer_size);
+  size_t sizeAll = 0;
+
+  for (int i = 0; i < initializer_size; ++i) {
+    const auto& tensor = model_proto_.graph().initializer(i);
+    const auto& name = tensor.name();
+    if (Contains(skipped_initializers_, name))
+      continue;
+
+    Shape shape;
+    for (auto dim : tensor.dims()) {
+      shape.push_back(SafeInt<uint32_t>(dim));
+    }
+
+    Type type = Type::TENSOR_FLOAT32;
+    switch (tensor.data_type()) {
+      case ONNX_NAMESPACE::TensorProto_DataType_FLOAT:
+        type = Type::TENSOR_FLOAT32;
+        break;
+      default:
+        // TODO: support other type
+        ORT_THROW("The initializer of graph doesn't have valid type, name: " +
+                  name + " type: " + std::to_string(tensor.data_type()));
+        break;
```

Comment:
Can we update this method to return a Status and exit gracefully in this case?

Reply:
This will be addressed with the above check if initializer is external as a TODO item (#8 in the PR descriptions)

Reply:
This seems orthogonal to that change. Couldn't you still have types that aren't supported by NNAPI? If you throw or use ORT_ENFORCE model execution fails. I would have thought we could fallback to cpu, so throwing seems incorrect to me.

---
In reply to: [444572387](https://github.com/microsoft/onnxruntime/pull/4287#discussion_r444572387) [](ancestors = 444572387)

Reply:
After the #8 todo in the description, since all the initializers used by the subgraph will be checked in GetCapability(), this ORT_THROW can be removed since we will never reach here, for now I would keep it this way, since later we may not need to return status at all because most of the ort_throw is related to input type which will be removed once we check each operator's input/initializers before build the model